File: CharAnalysis_2_0_MATLAB/CharPretreatment.m
function [Charcoal, Pretreatment, gapIn] = ...
        CharPretreatment(charData, site, Pretreatment, Results, plotData)
% CharPretreatment    Interpolate record, derive CHAR, optionally transform.
%   [Charcoal, Pretreatment, gapIn] =
%       CharPretreatment(charData, site, Pretreatment, Results, plotData)
%
%   Interpolates charcoal data to the resolution defined by yrInterp,
%   derives raw and resampled charcoal accumulation rates (CHAR), and
%   log-transforms CHAR if selected.
%
%   INPUTS
%     charData     : raw data matrix, Charster format
%                    cols: cmTop cmBot ageTop ageBot vol count
%     site         : site name string (for plot titles)
%     Pretreatment : struct  (zoneDiv, yrInterp, transform)
%     Results      : struct  (allFigures)
%     plotData     : scalar flag  1 = generate diagnostic plots
%                    (explicit argument in v2.0; was global in v1.1)
%
%   OUTPUTS
%     Charcoal     : struct with fields:
%                      cm, count, vol, con, ybp   (raw series)
%                      cmI, countI, volI, conI,
%                      accI, ybpI                  (resampled series)
%                      acc                         (raw CHAR)
%     Pretreatment : returned with yrInterp updated if it was 0 (auto)
%     gapIn        : [nGaps x 2] index matrix for missing-value gaps,
%                    or NaN(0,1) if there are no gaps
%
%   v2.0 changes vs v1.1
%     - plotData passed as explicit argument; global removed.
%     - nGaps defined unconditionally before the gap-interpolation block
%       so the plot section at the end never crashes when there are no
%       missing values (was an undefined-variable bug in v1.1).
%     - Double for-loop building propMatrix replaced by fully vectorised
%       broadcasting — four logical masks cover the four overlap cases.
%       Produces bit-identical results to the loop within floating-point
%       tolerance (~1e-14); benchmark before deploying on very large records.
%
%   CharAnalysis v2.0  -  Phase 1 modernisation

%% ── LOCAL VARIABLES ───────────────────────────────────────────────────────
zoneDiv     = Pretreatment.zoneDiv;
yrInterp    = Pretreatment.yrInterp;
transform   = Pretreatment.transform;
figPosition = [0.1343  0.2383  0.8648  0.6943];

%% ── TRIM RECORD TO zoneDiv BOUNDS ────────────────────────────────────────
if length(zoneDiv) > 1
    ybpStop  = zoneDiv(1);
    ybpStart = zoneDiv(end);
    charData = charData( charData(:,3) >= ybpStop & ...
                         charData(:,4) <= ybpStart, : );
end

%% ── SCREEN FOR MISSING VALUES (sample volume <= 0) ───────────────────────
%
%   nGaps is defined here unconditionally so the plot block at the bottom
%   of the function is always safe, even when nMissingValues == 0.

missingIdx     = find(charData(:,5) <= 0);
nMissingValues = length(missingIdx);
nGaps          = 0;            % defined unconditionally (v2.0 bug fix)
gapIn          = NaN(0,1);     % default: no gaps

if nMissingValues > 0

    % Identify contiguous runs of missing rows
    startIn        = missingIdx( [99; diff(missingIdx)] > 1 );
    endIn          = missingIdx( diff(missingIdx)        > 1 );
    endIn(end+1,1) = missingIdx(end);
    gapIn          = [startIn  endIn];
    nGaps          = size(gapIn, 1);

    cmGaps = sum( charData(gapIn(:,2)+1, 1) - charData(gapIn(:,1)-1, 1) );
    yrGaps = sum( charData(gapIn(:,2)+1, 3) - charData(gapIn(:,1)-1, 3) );

    disp(['NOTE: ' num2str(nMissingValues) ' missing value(s); ' ...
          num2str(nGaps) ' gap(s) totalling ' ...
          num2str(cmGaps) ' cm and ' num2str(round(yrGaps)) ' years.'])
    disp('      Values created via interpolation.')

    for i = 1:nGaps
        x        = [ charData(gapIn(i,1)-1, 1)
                     charData(gapIn(i,2)+1, 1) ];
        cmInterp = diff(charData(gapIn(i,1), 1:2));
        xi       = ( charData(gapIn(i,1)-1) : cmInterp : ...
                     charData(gapIn(i,2)+1, 1) )';
        y_vol    = [ charData(gapIn(i,1)-1, 5)
                     charData(gapIn(i,2)+1, 5) ];
        y_cnt    = [ charData(gapIn(i,1)-1, 6)
                     charData(gapIn(i,2)+1, 6) ];
        yi_vol   = interp1(x, y_vol, xi);
        yi_cnt   = interp1(x, y_cnt, xi);

        % Trim if interpolated length exceeds gap length
        % (can happen with variable sampling intervals around the gap)
        if length(yi_vol(2:end-1)) ~= length(gapIn(i,1):gapIn(i,2))
            yi_vol(2) = [];
            yi_cnt(2) = [];
        end

        charData(gapIn(i,1):gapIn(i,2), 5) = yi_vol(2:end-1);
        charData(gapIn(i,1):gapIn(i,2), 6) = yi_cnt(2:end-1);
    end
end

% Warn about non-contiguous samples not flagged as missing values
remainingGap = find( charData(2:end,1) - charData(1:end-1,2) );
if ~isempty(remainingGap)
    remCm = charData(remainingGap+1, 1) - charData(remainingGap,   1);
    remYr = charData(remainingGap+1, 3) - charData(remainingGap,   3);
    disp(['NOTE: ' num2str(length(remainingGap)) ' gap(s) in record ' ...
          'totalling ' num2str(sum(remCm)) ' cm and ' ...
          num2str(sum(remYr)) ' yr.'])
    disp('      Resampling will occur across this gap; consider filling')
    disp('      missing rows with -999 to interpolate over it instead.')
end

%% ── RETRIEVE VARIABLES FROM DATA MATRIX ──────────────────────────────────
Charcoal.cm    = charData(:,1);
Charcoal.count = charData(:,6);
Charcoal.vol   = charData(:,5);
Charcoal.con   = Charcoal.count ./ Charcoal.vol;
Charcoal.ybp   = charData(:,3);

%% ── SEDIMENT ACCUMULATION RATE ───────────────────────────────────────────
sedAcc = zeros(length(Charcoal.cm), 1);
for i = 1:length(Charcoal.ybp)-1
    sedAcc(i) = ( Charcoal.cm(i+1)  - Charcoal.cm(i) ) / ...
                ( Charcoal.ybp(i+1) - Charcoal.ybp(i) );
end

%% ── AUTO yrInterp IF NOT USER-DEFINED ────────────────────────────────────
if Pretreatment.yrInterp == 0
    yrInterp              = round(median(diff(charData(:,3))));
    Pretreatment.yrInterp = yrInterp;
    disp(' ')
    disp(['Record resampled to median resolution of ' ...
          num2str(Pretreatment.yrInterp) ' years.'])
end

%% ── BUILD RESAMPLED AGE VECTOR ───────────────────────────────────────────
Charcoal.ybpI = ( Pretreatment.zoneDiv(1) : Pretreatment.yrInterp : ...
                   Pretreatment.zoneDiv(end) )';

ageTop = charData(:,3);
ageBot = charData(:,4);
N_rs   = length(Charcoal.ybpI);

%% ── VECTORISED PROPORTION MATRIX ─────────────────────────────────────────
%
%   v1.1 used a double for-loop (up to ~500 000 iterations for a typical
%   record) to fill propMatrix(i,j), the fraction of raw sample j that
%   falls within resampled interval i.
%
%   v2.0 replaces this with a single vectorised operation.  Broadcasting
%   expands the [N_rs x 1] resampled-interval vectors against the
%   [1 x N_raw] raw-sample vectors to produce [N_rs x N_raw] logical masks
%   covering the four possible overlap geometries.
%
%   Results are numerically identical to the loop within floating-point
%   rounding (~1e-14 relative difference); confirm on your own datasets
%   with the original before retiring the loop version.

rsAgeTop = Charcoal.ybpI;                           % [N_rs x 1]
rsAgeBot = rsAgeTop + Pretreatment.yrInterp;        % [N_rs x 1]
aT       = ageTop';                                 % [1 x N_raw]
aB       = ageBot';                                 % [1 x N_raw]

% Case A: raw sample straddles the bottom edge of the resampled interval
%         overlap = rsAgeBot - ageTop
caseA = (aT >= rsAgeTop) & (aT < rsAgeBot) & (aB > rsAgeBot);

% Case B: raw sample straddles the top edge of the resampled interval
%         overlap = ageBot - rsAgeTop
caseB = (aT <  rsAgeTop) & (aB <= rsAgeBot) & (aB > rsAgeTop);

% Case C: raw sample lies entirely within the resampled interval
%         overlap = ageBot - ageTop
caseC = (aT >= rsAgeTop) & (aB <= rsAgeBot);

% Case D: resampled interval lies entirely within the raw sample
%         overlap = yrInterp (full interval)
caseD = (aT <  rsAgeTop) & (aB > rsAgeBot);

propMatrix = caseA .* (rsAgeBot - aT)              + ...
             caseB .* (aB       - rsAgeTop)         + ...
             caseC .* (aB       - aT)               + ...
             caseD .* Pretreatment.yrInterp;

propMatrix = propMatrix ./ Pretreatment.yrInterp;   % [N_rs x N_raw]

%% ── DERIVE RESAMPLED VALUES ───────────────────────────────────────────────
Charcoal.countI = NaN(N_rs, 1);
Charcoal.volI   = NaN(N_rs, 1);
Charcoal.conI   = NaN(N_rs, 1);
sedAccI         = NaN(N_rs, 1);

for i = 1:N_rs
    in = propMatrix(i,:) > 0;
    if any(in)
        Charcoal.countI(i) = sum( Charcoal.count(in) .* propMatrix(i,in)' );
        Charcoal.volI(i)   = sum( Charcoal.vol(in)   .* propMatrix(i,in)' );
        Charcoal.conI(i)   = sum( Charcoal.con(in)   .* propMatrix(i,in)' );
        sedAccI(i)         = sum( sedAcc(in)          .* propMatrix(i,in)' );
    end
end

% Resampled depths via linear interpolation (unchanged from v1.1)
Charcoal.cmI = interp1(Charcoal.ybp, Charcoal.cm, Charcoal.ybpI);

%% ── CHARCOAL ACCUMULATION RATES ──────────────────────────────────────────
Charcoal.acc  = ( Charcoal.con  .* sedAcc )';   % raw CHAR
Charcoal.accI = ( Charcoal.conI .* sedAccI );   % resampled CHAR

%% ── TRANSFORM IF SELECTED ────────────────────────────────────────────────
if transform == 1
    Charcoal.accI = log10(Charcoal.accI + 1);
end
if transform == 2
    Charcoal.accI = log(Charcoal.accI + 1);
end

%% ── PLOT RAW AND RESAMPLED SERIES (FIGURE 1, SUBPLOT 1) ─────────────────
%
%   Only drawn when plotData == 1 and Results.allFigures == 1.
%   nGaps is always defined at this point (v2.0 bug fix).

if plotData == 1 && Results.allFigures == 1

    figure(1); clf;
    set(gcf, 'name', ...
        '(a) C_raw and C_resampled;  (b) C_raw and options for C_background', ...
        'units', 'normalized', 'position', figPosition, 'color', 'w')

    subplot(2,1,1)
    h = bar(Charcoal.ybp, Charcoal.acc, 1);
    set(h, 'facecolor', [.5 .5 .5], 'edgecolor', [.5 .5 .5])
    ylim([0, prctile(Charcoal.acc, 99)]);
    hold on
    stairs(Charcoal.ybpI - 0.5*Pretreatment.yrInterp, Charcoal.accI, ...
           '.-k', 'linewidth', 1)

    if nGaps > 0
        for i = 1:nGaps
            yax   = get(gca, 'ylim');
            xpatch = [ Charcoal.ybp(gapIn(i,1))  Charcoal.ybp(gapIn(i,1)) ...
                       Charcoal.ybp(gapIn(i,2))  Charcoal.ybp(gapIn(i,2)) ];
            patch(xpatch, [yax(1) yax(2) yax(2) yax(1)], 'w', ...
                  'edgecolor', [.75 .75 .75], 'facecolor', [.75 .75 .75])
        end
        legend('C_r_a_w', ...
               ['C_i_n_t_e_r_p_o_l_a_t_e_d: ' num2str(yrInterp) ' yr'], ...
               'missing values')
    else
        legend('C_r_a_w', ...
               ['C_i_n_t_e_r_p_o_l_a_t_e_d: ' num2str(yrInterp) ' yr'])
    end

    xlabel('time (cal. yr BP)')
    ylabel('CHAR (# cm^-^2 yr^-^1)')
    set(gca, 'xlim',    [min(Charcoal.ybp)-100  max(Charcoal.ybp)], ...
             'box',     'off', ...
             'tickdir', 'out', ...
             'xdir',    'rev')
    title([{char(site)}, {'(a) C_r_a_w and C_i_n_t_e_r_p_o_l_a_t_e_d'}])

end

end

File: CharAnalysis_2_0_R/R/CharPretreatment.R
#' Interpolate and pretreat a charcoal time series
#'
#' Resamples the raw charcoal data to equal time steps using a
#' proportion-weighted scheme, fills missing-value gaps by linear
#' interpolation, computes charcoal accumulation rates (CHAR), and optionally
#' applies a log transformation.
#'
#' Mirrors \code{CharPretreatment.m} from the MATLAB v2.0 codebase.  The
#' vectorised proportion matrix (four broadcast cases) produces results
#' numerically identical to the MATLAB implementation within floating-point
#' tolerance (~1e-14).
#'
#' @param char_data    Numeric matrix (n x 6+): cmTop, cmBot, ageTop, ageBot,
#'   charVol, charCount. Rows sorted youngest-first (ascending ageTop).
#' @param site         Character string; site name used in diagnostic messages.
#' @param pretreatment Named list with elements:
#'   \describe{
#'     \item{zoneDiv}{Numeric vector of zone boundaries (cal. yr BP),
#'       strictly ascending, at least 2 values.}
#'     \item{yrInterp}{Resampling interval (yr).  0 = use median raw
#'       resolution (auto).}
#'     \item{transform}{Integer: 0 = none; 1 = log10(x+1); 2 = ln(x+1).}
#'   }
#' @param results      Named list; only \code{allFigures} is referenced.
#'   Pass \code{NULL} to suppress any plot-related behaviour (no plots are
#'   produced in the R implementation in Phase 1).
#' @param plot_data    0/1 integer flag.  Ignored in R (no diagnostic plots);
#'   included for API symmetry with the MATLAB function signature.
#'
#' @return Named list with three elements:
#'   \describe{
#'     \item{charcoal}{List of raw and resampled series:
#'       \code{cm, count, vol, con, ybp, acc} (raw);
#'       \code{cmI, countI, volI, conI, accI, ybpI} (resampled).}
#'     \item{pretreatment}{Input list returned with \code{yrInterp} updated
#'       when it was 0 (auto), and \code{zoneDiv[end]} corrected if it
#'       exceeded the bottom age of the last raw sample.}
#'     \item{gap_in}{Integer matrix (nGaps x 2) of gap row-index pairs, or
#'       \code{matrix(NA_integer_, 0, 2)} when no gaps exist.}
#'   }
#'
#' @details
#'   ## Proportion matrix
#'   For each resampled interval \eqn{i} and each raw sample \eqn{j},
#'   \code{prop_matrix[i,j]} is the fraction of the raw sample's age span
#'   that falls within the resampled interval.  Four mutually exclusive
#'   overlap geometries (Cases A-D) are evaluated via matrix broadcasting
#'   across the full \eqn{[N_{rs} \times N_{raw}]} grid -- no loops required.
#'
#'   | Case | Geometry                                      | Overlap            |
#'   |------|-----------------------------------------------|--------------------|
#'   | A    | Raw straddles the **bottom** edge             | rsAgeBot - ageTop  |
#'   | B    | Raw straddles the **top** edge                | ageBot - rsAgeTop  |
#'   | C    | Raw lies **entirely within** resampled interval | ageBot - ageTop  |
#'   | D    | Resampled lies **entirely within** raw sample | yrInterp           |
#'
#'   ## zoneDiv auto-correction
#'   If \code{zoneDiv[end]} exceeds the bottom age of the last raw sample,
#'   the terminal resampled intervals would have no overlapping raw data and
#'   \code{accI} would be \code{NA}.  These \code{NA}s propagate into
#'   \code{charBkg} and can hang the GMM in Phase 2.  The value is silently
#'   corrected to \code{lastAgeBotInData} and the user is notified
#'   (v2.0 behaviour, preserved here).
#'
#' @seealso [char_parameters()], [char_validate_params()], [CharAnalysis()]
char_pretreatment <- function(char_data, site, pretreatment,
                               results = NULL, plot_data = 1L) {

  zone_div  <- pretreatment$zoneDiv
  yr_interp <- pretreatment$yrInterp
  transform <- pretreatment$transform

  # =========================================================================
  # TRIM RECORD TO zoneDiv BOUNDS
  # =========================================================================
  if (length(zone_div) > 1L) {
    ybp_stop  <- zone_div[1L]
    ybp_start <- zone_div[length(zone_div)]
    keep      <- char_data[, 3L] >= ybp_stop & char_data[, 4L] <= ybp_start
    char_data <- char_data[keep, , drop = FALSE]
  }

  # =========================================================================
  # AUTO-CORRECT zoneDiv[end] IF IT EXCEEDS THE DATA BOUNDARY
  #
  #   If zoneDiv[end] extends beyond the bottom age of the last raw sample,
  #   the vectorised proportion matrix assigns NA to those terminal resampled
  #   intervals.  These NAs propagate into charBkg and can hang the GMM.
  #   Correct silently and notify the user (v2.0 behaviour).
  # =========================================================================
  last_age_bot <- char_data[nrow(char_data), 4L]
  n_zones      <- length(pretreatment$zoneDiv)

  if (pretreatment$zoneDiv[n_zones] > last_age_bot) {
    message("NOTE: zoneDiv[end] (", pretreatment$zoneDiv[n_zones],
            " yr BP) exceeds the bottom age of the last raw sample (",
            last_age_bot, " yr BP). zoneDiv[end] corrected to ",
            last_age_bot, " yr BP.")
    pretreatment$zoneDiv[n_zones] <- last_age_bot
    zone_div <- pretreatment$zoneDiv
  }

  # =========================================================================
  # SCREEN FOR MISSING VALUES (sample volume <= 0)
  #
  #   nGaps is defined unconditionally here so it is always safe to reference
  #   (v2.0 bug fix -- in v1.1 it was undefined when nMissingValues == 0).
  # =========================================================================
  missing_idx  <- which(char_data[, 5L] <= 0)
  n_missing    <- length(missing_idx)
  n_gaps       <- 0L
  gap_in       <- matrix(NA_integer_, nrow = 0L, ncol = 2L)

  if (n_missing > 0L) {

    # Identify contiguous runs: a new run starts whenever consecutive missing
    # indices are non-adjacent (diff > 1).  Prepending 99 makes the first
    # missing index always qualify as a run start (99 > 1 is TRUE).
    diff_idx <- c(99L, diff(missing_idx))
    start_in <- missing_idx[diff_idx > 1L]

    # A run ends at positions where the jump to the next missing index > 1,
    # plus always at the very last missing index.
    diff_end <- diff(missing_idx)
    end_in   <- c(missing_idx[diff_end > 1L], missing_idx[n_missing])

    gap_in <- cbind(start_in, end_in)
    n_gaps <- nrow(gap_in)

    cm_gaps <- sum(char_data[gap_in[, 2L] + 1L, 1L] -
                   char_data[gap_in[, 1L] - 1L, 1L])
    yr_gaps <- sum(char_data[gap_in[, 2L] + 1L, 3L] -
                   char_data[gap_in[, 1L] - 1L, 3L])

    message("NOTE: ", n_missing, " missing value(s); ", n_gaps,
            " gap(s) totalling ", cm_gaps, " cm and ",
            round(yr_gaps), " years.")
    message("      Values created via interpolation.")

    for (g in seq_len(n_gaps)) {
      i1 <- gap_in[g, 1L]
      i2 <- gap_in[g, 2L]

      # Anchor depths and volumes from the rows surrounding the gap
      x_anchor <- c(char_data[i1 - 1L, 1L], char_data[i2 + 1L, 1L])
      cm_step  <- diff(char_data[i1, 1:2])
      xi       <- seq(char_data[i1 - 1L, 1L], char_data[i2 + 1L, 1L],
                      by = cm_step)

      y_vol <- c(char_data[i1 - 1L, 5L], char_data[i2 + 1L, 5L])
      y_cnt <- c(char_data[i1 - 1L, 6L], char_data[i2 + 1L, 6L])

      yi_vol <- approx(x_anchor, y_vol, xout = xi)$y
      yi_cnt <- approx(x_anchor, y_cnt, xout = xi)$y

      # Trim if the interpolated sequence is longer than the gap
      # (can happen with variable sampling intervals around the gap;
      # mirrors the MATLAB trim logic exactly)
      gap_len <- length(i1:i2)
      inner_vol <- yi_vol[-c(1L, length(yi_vol))]
      inner_cnt <- yi_cnt[-c(1L, length(yi_cnt))]
      if (length(inner_vol) != gap_len) {
        inner_vol <- inner_vol[-2L]
        inner_cnt <- inner_cnt[-2L]
      }

      char_data[i1:i2, 5L] <- inner_vol
      char_data[i1:i2, 6L] <- inner_cnt
    }
  }

  # Non-contiguous sample warning (informational only; does not affect output)
  remaining_gap <- which((char_data[-1L, 1L] -
                           char_data[-nrow(char_data), 2L]) != 0)
  if (length(remaining_gap) > 0L) {
    rem_cm <- sum(char_data[remaining_gap + 1L, 1L] -
                  char_data[remaining_gap, 1L])
    rem_yr <- sum(char_data[remaining_gap + 1L, 3L] -
                  char_data[remaining_gap, 3L])
    message("NOTE: ", length(remaining_gap), " gap(s) in record totalling ",
            rem_cm, " cm and ", rem_yr, " yr.")
    message("      Resampling will occur across this gap; consider filling")
    message("      missing rows with -999 to interpolate over it instead.")
  }

  # =========================================================================
  # EXTRACT RAW SERIES
  # =========================================================================
  charcoal <- list(
    cm    = char_data[, 1L],
    count = char_data[, 6L],
    vol   = char_data[, 5L],
    con   = char_data[, 6L] / char_data[, 5L],
    ybp   = char_data[, 3L]
  )

  # =========================================================================
  # SEDIMENT ACCUMULATION RATE
  #
  #   Mirrors MATLAB: sedAcc = zeros(n,1); loop fills positions 1:(n-1).
  #   The last element therefore stays 0, which is reproduced here via
  #   c(diff/diff, 0).
  # =========================================================================
  sed_acc <- c(diff(charcoal$cm) / diff(charcoal$ybp), 0)

  # =========================================================================
  # AUTO yrInterp (if not user-specified)
  # =========================================================================
  if (pretreatment$yrInterp == 0) {
    yr_interp            <- round(median(diff(char_data[, 3L])))
    pretreatment$yrInterp <- yr_interp
    message("\nRecord resampled to median resolution of ",
            pretreatment$yrInterp, " years.")
  }

  # =========================================================================
  # BUILD RESAMPLED AGE VECTOR
  #
  #   Mirrors MATLAB colon operator:
  #     zoneDiv(1) : yrInterp : zoneDiv(end)
  # =========================================================================
  charcoal$ybpI <- seq(from = pretreatment$zoneDiv[1L],
                       to   = pretreatment$zoneDiv[length(pretreatment$zoneDiv)],
                       by   = pretreatment$yrInterp)

  age_top <- char_data[, 3L]
  age_bot <- char_data[, 4L]
  N_rs    <- length(charcoal$ybpI)
  N_raw   <- length(age_top)

  rs_age_top <- charcoal$ybpI
  rs_age_bot <- rs_age_top + pretreatment$yrInterp

  # =========================================================================
  # VECTORISED PROPORTION MATRIX  [N_rs x N_raw]
  #
  #   For each pair (resampled interval i, raw sample j), prop_matrix[i,j]
  #   is the fraction of raw sample j's age span that falls within the
  #   resampled interval i.
  #
  #   Broadcasting: both rs_* vectors ([N_rs]) and age_* vectors ([N_raw])
  #   are expanded to [N_rs x N_raw] matrices before comparison.
  #
  #     rsT[i,j] = rs_age_top[i]   (same value across each row)
  #     rsB[i,j] = rs_age_bot[i]
  #     aT[i,j]  = age_top[j]      (same value down each column)
  #     aB[i,j]  = age_bot[j]
  #
  #   Case A: raw sample straddles the BOTTOM edge of the resampled interval
  #           (ageTop inside, ageBot outside below)
  #           overlap = rsAgeBot[i] - ageTop[j]
  #
  #   Case B: raw sample straddles the TOP edge of the resampled interval
  #           (ageBot inside, ageTop outside above)
  #           overlap = ageBot[j] - rsAgeTop[i]
  #
  #   Case C: raw sample lies ENTIRELY WITHIN the resampled interval
  #           overlap = ageBot[j] - ageTop[j]
  #
  #   Case D: resampled interval lies ENTIRELY WITHIN the raw sample
  #           overlap = yrInterp
  #
  #   Note: Cases C and A share (aT >= rsT); Case C is the stricter subset
  #   (aB <= rsB), while Case A requires (aB > rsB).  The four cases are
  #   mutually exclusive and exhaustive for all overlapping pairs.
  # =========================================================================

  # Expand to [N_rs x N_raw] matrices
  rsT <- matrix(rs_age_top, nrow = N_rs, ncol = N_raw)            # rs_age_top[i]
  rsB <- matrix(rs_age_bot, nrow = N_rs, ncol = N_raw)            # rs_age_bot[i]
  aT  <- matrix(age_top,   nrow = N_rs, ncol = N_raw, byrow = TRUE) # age_top[j]
  aB  <- matrix(age_bot,   nrow = N_rs, ncol = N_raw, byrow = TRUE) # age_bot[j]

  caseA <- (aT >= rsT) & (aT  < rsB) & (aB >  rsB)   # straddles bottom edge
  caseB <- (aT  < rsT) & (aB <= rsB) & (aB >  rsT)   # straddles top edge
  caseC <- (aT >= rsT) & (aB <= rsB)                  # entirely within
  caseD <- (aT  < rsT) & (aB  > rsB)                  # raw contains interval

  prop_matrix <- caseA * (rsB - aT)              +
                 caseB * (aB  - rsT)             +
                 caseC * (aB  - aT)              +
                 caseD *  pretreatment$yrInterp

  prop_matrix <- prop_matrix / pretreatment$yrInterp   # normalise to fraction

  # =========================================================================
  # DERIVE RESAMPLED VALUES
  #
  #   For each resampled interval i, sum the contributions from all raw
  #   samples j where prop_matrix[i,j] > 0.  This mirrors the MATLAB loop:
  #     in = propMatrix(i,:) > 0;
  #     Charcoal.countI(i) = sum( count(in) .* propMatrix(i,in)' );
  # =========================================================================
  charcoal$countI <- rep(NA_real_, N_rs)
  charcoal$volI   <- rep(NA_real_, N_rs)
  charcoal$conI   <- rep(NA_real_, N_rs)
  sed_acc_I       <- rep(NA_real_, N_rs)

  for (i in seq_len(N_rs)) {
    in_idx <- prop_matrix[i, ] > 0
    if (any(in_idx)) {
      p <- prop_matrix[i, in_idx]
      charcoal$countI[i] <- sum(charcoal$count[in_idx] * p)
      charcoal$volI[i]   <- sum(charcoal$vol[in_idx]   * p)
      charcoal$conI[i]   <- sum(charcoal$con[in_idx]   * p)
      sed_acc_I[i]       <- sum(sed_acc[in_idx]          * p)
    }
  }

  # Resampled depths: linear interpolation of raw cm vs. age.
  # rule = 1: return NA outside the data range, matching MATLAB's interp1
  # default (which returns NaN for out-of-range queries).
  charcoal$cmI <- approx(charcoal$ybp, charcoal$cm,
                          xout = charcoal$ybpI, rule = 1L)$y

  # =========================================================================
  # CHARCOAL ACCUMULATION RATES
  # =========================================================================
  charcoal$acc  <- charcoal$con  * sed_acc     # raw CHAR (# cm-2 yr-1)
  charcoal$accI <- charcoal$conI * sed_acc_I   # resampled CHAR

  # =========================================================================
  # OPTIONAL LOG TRANSFORM
  # =========================================================================
  if (transform == 1L) {
    charcoal$accI <- log10(charcoal$accI + 1)
  }
  if (transform == 2L) {
    charcoal$accI <- log(charcoal$accI + 1)
  }

  # =========================================================================
  # RETURN
  # =========================================================================
  list(
    charcoal     = charcoal,
    pretreatment = pretreatment,
    gap_in       = gap_in
  )
}
